io/pci/PowerNVEEH.py

```diff
@@ -65,13 +65,21 @@ def setUp(self):
         cmd = "echo %d > /sys/kernel/debug/powerpc/eeh_max_freezes"\
             % self.max_freeze
         process.system(cmd, ignore_status=True, shell=True)
-        self.function = str(self.params.get('function'))
-        self.err = str(self.params.get('err'))
+        self.function = str(self.params.get('function'), default='4')
         self.pci_device = str(self.params.get('pci_device', default=' '))
         self.phb = self.pci_device.split(":", 1)[0]
         self.addr = genio.read_file("/sys/bus/pci/devices/%s/"
                                     "eeh_pe_config_addr" % self.pci_device)
         self.addr = str(self.addr).rstrip()
+        self.err = 0
```

Comment:
Its better not to assume this. We should be doing a check, may add else condition below, after checking for 64bit

Reply:
This is not assumption. This is initialising to default value. Only in one case, it is 1, which is handled in `if`

Reply:
I meant, what if lspci itself was reporting wrong, in that case we miss it. Also, there is good chance there could be 128 bit or something like that.

Reply:
@vrbagalkote What alternatives do we have ?

Reply:
```
   if 'Memory' in line and '64-bit, prefetchable' in line:
                self.err = 1
   else 'Memory' in line and '32-bit, ...' in line:
                self.err = 0
```
is better

Reply:
https://en.wikipedia.org/wiki/PCI_configuration_space

From here, only `32 and 64 bit`, and only `prefetchable and non-prefetchable` are possible.

So, out of 4 cases, only 1 case needs `err` value to be 1. So, the initial approach is better here :-)

Reply:
Two things:

1: From Gavin's note, error code can be 0/1 for any adapter, its only based on function we are going to use while injecting error. Its not dependent on the adapter configuration size.

2: From the notes you shared, also says for direct PHB adapters, address and mask should be 0, 0. You are not checking the adapter whether its direct PHB or not. So I am sceptical whether this change is required at the first place.
```
Gavin's Note:

In the early days of EEH development, the debugfs entry and command are used to
to error injection (Frozen PE case). However, it's not reliable way. At the meanwhile,
I plan to drop the debugfs entry pretty soon. Instead, the below debugfs entry is the
correct one for error injection. 

    /sys/kernel/debug/powerpc/PCIxxxx/err_injct

The debugfs entry accepts commands "PE_number:<0,1>:<function>:0:0"

PE_number: The number of PE the specified PCI device is attached to. It can be retrieved
           from /sys/bus/pci/devices/xxxx:xx:xx.x/eeh_pe_config_addr
<0,1>:     0 - 32-bits error injection. When doing error injection via PCI config, IO or
           32-bits MMIO access.
           1 - 64-bits error injection. When doing error injection via 64-bits MMIO access
           or DMA.
<function>: Please refer to Linux source code arch/powerpc/include/asm/opal-api.h::OpalErrinjectFunc
```
So, finally error code 0/1 is not dependent on adapter configuration. We should be covering both errors on any adapter, thats the reason why its in yaml configuration file. So its up to user.

Reply:
1. Error being 0 / 1, is based on lspci output. Can you try once injecting error 0 for 64-bit prefetchable BARs and let me know ?

Reply:
The statement I pasted is from commit log message. And guliherme's notes is for his reference, and he put a clear disclaimer. :). So I would like to believe commit message than a personal reference notes

Reply:
Ok, are we clear about direct PHB slots ?

Reply:
About error being 0/1, can you please point me exactly where it says it is user configurable and not based on PCI config space ?

Reply:
@vrbagalkote I tried the script with same patches on Power8 server, on a direct PHB slot:

Reply:
Thanks @narasimhan-v . You have cleared one ambiguity out of the way. So, we can use address mask and memory for any kind of adapter. But having said that we can also 0 and 0 also for any adapter. So I am happy either way. 

One last thing I want to is, Can you also check if error code 0 and 1 works fine for any adapter?
If it works, then there is no need to calculate it in the script, it can be given via yaml file(user configurable).

Reply:
> So I am happy either way.

Cool

Reply:
Exactly for this reason, let user decide which function and which error type. There is a chance that user gives MMIO function and scripts evaluates to wrong error code. Let user decide both, till we figure out what exactly.

Reply:
@vrbagalkote For MMIO function, we should calculate from lspci. So, it should be computed.

Reply:
>  There is a chance that user gives MMIO function and scripts evaluates to wrong error code.